### PR TITLE
feature/21648-upper-and-lower-whisker-width

### LINKED ISSUE #21648: lowerWhiskerWidth/upperWhiskerWidth property
# Description of the feature
A property that can hide or show the lower or upper whisker. It may also function as a simple lowerWhiskerWidth/upperWhiskerWidth property.

# Library related to the feature
Highcharts, boxplot series

# Proof of Concept/Live example for the feature
Demo with SVG Renderer: https://jsfiddle.net/BlackLabel/bgnv2dxs/
image

----

**You can vote for this feature by adding a thumbs-up reaction to this post.**

### PR DESCRIPTION
**Pull request by MarkusBarstad**
Added support for individual whisker length for box plots and error bars. See #21648.

---
~~Closes #21648, Optional length for each whisker individually.~~

- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208031949890257

### CODE REVIEW

**Comment by khlieng** [comment]
Fed it to o1 🤷

### Overall Review:

This pull request introduces the ability to configure custom upper and lower whisker lengths individually for BoxPlot series in Highcharts. It modifies the BoxPlotSeries implementation to handle separate `upperWhiskerLength` and `lowerWhiskerLength`, updates TypeScript definitions to include these new options, and adds a demo and unit tests to showcase and test the new functionality.

This is a valuable enhancement that provides more flexibility for users who need to adjust whisker lengths individually.

### Detailed Feedback:

#### 1. `samples/highcharts/series-boxplot/custom-whisker-length/*`

- **Files**: `demo.css`, `demo.details`, `demo.html`, `demo.js`
- **Review**:
  - The new sample demonstrates how to use the custom whisker lengths effectively.
  - The demo code is clean and well-documented.
  - **Suggestion**: In `demo.js`, the chart title has a typo:

    ```javascript
    title: {
        text: 'Configuring whisker length's'
    },
    ```

    The apostrophe in `length's` should be removed:

    ```javascript
    title: {
        text: 'Configuring whisker lengths'
    },
    ```

#### 2. `samples/unit-tests/series-boxplot/boxplot/demo.js`

- **Review**:
  - The added unit test for individual point options and `Point.update` is comprehensive.
  - It not only checks the rendering of colors and widths but also verifies that the custom upper and lower whisker lengths are applied correctly.
  - **Suggestion**: Consider adding comments to explain the purpose of each assertion for future maintainability.

#### 3. `ts/Series/BoxPlot/BoxPlotPoint.d.ts`

- **Review**:
  - The new properties `upperWhiskerLength` and `lowerWhiskerLength` are appropriately added to the `BoxPlotPoint` class.
  - **No issues found**.

#### 4. `ts/Series/BoxPlot/BoxPlotPointOptions.d.ts`

- **Review**:
  - The documentation for the new properties `upperWhiskerLength` and `lowerWhiskerLength` is clear and includes a reference to a sample.
  - **Suggestion**:
    - Make sure to mention default behaviors or values if applicable.
    - For consistency, ensure that the description language matches that of other properties in the file.

#### 5. `ts/Series/BoxPlot/BoxPlotSeries.ts`

- **Review**:
  - The code has been refactored to handle individual whisker lengths.
  - The use of `geWhiskerPair` function inside the loop is a good approach to keep the code organized.
  - **Issues and Suggestions**:
    - **Performance Concern**: The function `geWhiskerPair` is defined inside the loop over `points`. This means a new function is created on each iteration, which may have performance implications. It would be better to define this function outside the loop.

      ```typescript
      // Define the function outside the loop
      const getWhiskerPair = (point: BoxPlotPoint): SVGPath => { ... };

      for (const point of points) {
          // Use getWhiskerPair inside the loop
          point.whiskers[verb]({ d: getWhiskerPair(point) });
      }
      ```

    - **Typo in Function Name**: The function `geWhiskerPair` seems to be missing a 't' in 'get'. It should be `getWhiskerPair`.
    - **Variable Shadowing**: There is a redefinition of `halfWidth` inside the loop, which might lead to confusion. Ensure consistent use of variables.

#### 6. `ts/Series/BoxPlot/BoxPlotSeriesOptions.d.ts`

- **Review**:
  - The `whiskerLength` option has been updated, but there's a typo in the union type:

    ```typescript
    whiskerLength?: (
        BoxPlotPoint['upperWhiskerLength'] |
        BoxPlotPoint['upperWhiskerLength']
    );
    ```

    The second type in the union should be `BoxPlotPoint['lowerWhiskerLength']`:

    ```typescript
    whiskerLength?: (
        BoxPlotPoint['upperWhiskerLength'] |
        BoxPlotPoint['lowerWhiskerLength']
    );
    ```

  - **Suggestion**:
    - Update the documentation/comments for `whiskerLength` to reflect that it can now be overridden by `upperWhiskerLength` and `lowerWhiskerLength` at the point level.
    - Ensure that the descriptions are clear on how these properties interact.

### General Suggestions:

- **Consistency in Naming**: Ensure that all variable and function names are consistently spelled and follow the project's naming conventions.
- **Documentation**: Update the Highcharts API documentation to include the new options, and explain how they override the `whiskerLength` property.
- **Unit Tests**: Consider adding more unit tests to cover edge cases, such as when only one of the `upperWhiskerLength` or `lowerWhiskerLength` is set, or when invalid values are provided.
- **Type Checking**: Ensure that all new TypeScript types are correctly defined and that they won't cause issues in projects that consume Highcharts with strict type checking.

### Conclusion:

The pull request adds valuable functionality to the BoxPlot series and is mostly well-implemented. Addressing the minor issues and suggestions above will improve the quality and maintainability of the code.

Thank you for your contribution!

**Comment by MarkusBarstad** [comment]
> Fed it to o1 🤷

Well, that was actually some decent feedback @khlieng ! 🤣 

Esp. the typos and performance thing, I will fix!

File: ts/Series/BoxPlot/BoxPlotSeries.ts [hunk 1]
```diff
@@ -297,25 +294,62 @@ class BoxPlotSeries extends ColumnSeries {
...
+                                        number |
+                                        string |
+                                        undefined
+                                    ) = whiskerLength,
+                                    yPos: number
+                                ): SVGPath.Segment[] => {
+                                    const halfLen = (
+                                            (
+                                                typeof xLen === 'string' &&
+                                            /%$/.test(xLen)
```
Comment (goransle) [inline ts/Series/BoxPlot/BoxPlotSeries.ts:315]: Would prefer `xLen.endsWith('%')` for readability here

File: ts/Series/BoxPlot/BoxPlotSeries.ts [hunk 2]
```diff
@@ -297,25 +294,62 @@ class BoxPlotSeries extends ColumnSeries {
...
-                    ];
-                    point.whiskers[verb]({ d });
+                    // Moved here as "const", refactoring of whiskerlength
+                    // triggered "no-loop-func" from eslint
+                    const halfWidth = width / 2,
+                        geWhiskerPair = (
+                            point: BoxPlotPoint
+                        ): SVGPath => {
+                            const strokeWidth = point.whiskers.strokeWidth(),
+                                getWhisker = (
```
Comment (goransle) [inline ts/Series/BoxPlot/BoxPlotSeries.ts:304]: I would maybe consider drawing these functions out as class methods or static functions. Better for readability, and would make it easier to write unit tests :)

**Review by goransle — APPROVED** [review]
Looks good. Maybe merge in master though, to hopefully get the visual comparison 🤞

File: ts/Series/BoxPlot/BoxPlotSeries.ts [hunk 3]
```diff
@@ -297,25 +352,17 @@ class BoxPlotSeries extends ColumnSeries {
...
-                        (/%$/).test(whiskerLength)
-                    ) ?
-                        halfWidth * parseFloat(whiskerLength) / 100 :
-                        Number(whiskerLength) / 2;
-                    d = [
-                        // High whisker
-                        ['M', crisp(stemX - pointWiskerLength), highPlot],
-                        ['L', crisp(stemX + pointWiskerLength), highPlot],
+                    // Moved here as "const", refactoring of whiskerlength
+                    // triggered "no-loop-func" from eslint
```
Comment (TorsteinHonsi) [inline ts/Series/BoxPlot/BoxPlotSeries.ts:356]: This comment refers to a historic state that is no longer relevant for us in the future when we want to edit this code. It's better that comments explain the code as it currently works.

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks, great work! Nothing to improve on the code as far as I can see.

There's an inconsistency however that I think we must address. The existing `whiskerLength` property exists on the series level. But the new properties exist only on the point level and have no effect on the series level: https://jsfiddle.net/highcharts/xw25ubn3/

Contrary, the existing `whiskerLength` has no effect on individual points: https://jsfiddle.net/highcharts/xw25ubn3/1/

Is there a good logic reason for this distinction? If not, I think they should work the same, so that all three options can apply to both series and points. The more specific settings (lower and upper), would take precedence over the general `whiskerLength`.

**Comment by MarkusBarstad** [comment]
> Is there a good logic reason for this distinction? If not, I think they should work the same, so that all three options can apply to both series and points.

@TorsteinHonsi Good points! 

If I understand you correctly:

Props of points should override those from series _except_:
- `series.upperWhiskerLength` overrides `point.whiskerLength`, 
- Same goes for `series.lowerWhiskerLength`.

Implementation of this: https://jsfiddle.net/mkb93/otwqb047/

Regarding docs:
Currently Ive structured the DOCs so that each option informs about/links to any settings which will be overriden. 

Should we also have each option inform on settings which may supersede it?

**Comment by TorsteinHonsi** [comment]
> `series.upperWhiskerLength` overrides `point.whiskerLength`

Yes, I think that makes more sense. The idea is that the more specific one should override the more general one, but in this case it depends. In one way, _upper_ whisker length is more specific. In another way, _point_ whisker length is more specific than _series_. So I think we just need to decide on one and document it.

**Comment by MarkusBarstad** [comment]
Alright @TorsteinHonsi , every whisker option now mentions and links to any whisker option it might override/be overriden by.

Not sure if that's a too verbose approach but it did not feel right to leave anything out. 🤔

Also note:
The local API server won't run (issue with the `jsdoc` script) so I am not sure if the links in the proposed docs are functional.

File: ts/Series/BoxPlot/BoxPlotSeries.ts [hunk 4]
```diff
@@ -296,26 +356,25 @@ class BoxPlotSeries extends ColumnSeries {
...
-                        ['M', crisp(stemX - pointWiskerLength), highPlot],
-                        ['L', crisp(stemX + pointWiskerLength), highPlot],
+                if (pointWhiskerLength) {
+                    const halfWidth = width / 2,
+                        whiskers = this.getWhiskerPair(
+                            halfWidth,
+                            stemX,
+                            (
+                                point.upperWhiskerLength ||
+                                options.upperWhiskerLength ||
```
Comment (TorsteinHonsi) [inline ts/Series/BoxPlot/BoxPlotSeries.ts:366]: I think we should use `??` rather than `||` here. It is intuitive for an implementer to disable the whisker by setting the length to 0. See https://jsfiddle.net/highcharts/xw25ubn3/2/ . This is also how a `whiskerLength` of 0 behaves.

File: ts/Series/BoxPlot/BoxPlotSeries.ts [hunk 5]
```diff
@@ -296,26 +356,25 @@ class BoxPlotSeries extends ColumnSeries {
...
+                                pointWhiskerLength
+                            ),
+                            (
+                                point.lowerWhiskerLength ||
+                                options.lowerWhiskerLength ||
```
Comment (TorsteinHonsi) [inline ts/Series/BoxPlot/BoxPlotSeries.ts:371]: As above, we probably want `??`

File: ts/Series/BoxPlot/BoxPlotSeries.ts [hunk 6]
```diff
@@ -87,6 +87,65 @@ class BoxPlotSeries extends ColumnSeries {
...
+            ): SVGPath.Segment[] => {
+                const halfLen = (
+                        (
+                            typeof xLen === 'string' &&
+                            xLen.endsWith('%')
+                        ) ?
+                            (
+                                halfWidth * parseFloat(xLen)
+                            ) / 100 :
+                            Number(xLen) / 2
```
Comment (TorsteinHonsi) [inline ts/Series/BoxPlot/BoxPlotSeries.ts:116]: We probably want to use the `relativeLength` utility function that is designed exactly for this pattern.

File: samples/highcharts/series-boxplot/whisker-length/demo.details [hunk 7]
```diff
@@ -0,0 +1,6 @@
+---
+name: Box plot
```
Comment (TorsteinHonsi) [inline samples/highcharts/series-boxplot/whisker-length/demo.details:2]:
```suggestion
name: Box plot whisker lengths
```

File: samples/highcharts/series-boxplot/custom-whisker-length/demo.css
Comment (TorsteinHonsi) [inline samples/highcharts/series-boxplot/custom-whisker-length/demo.css:1]: The `custom-whisker-length` demo is not refered to at all in the docs, and seems to be all covered by the `whisker-length` demo. Did you mean to delete it?

File: samples/highcharts/series-boxplot/whisker-length/demo.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        title: false
+    },
+
+    legend: {
+        enabled: false
+    },
+
+    series: [{
+        whiskerLength: 100,
+        lowerWhiskerLength: 10,
```
Comment (TorsteinHonsi) [inline samples/highcharts/series-boxplot/whisker-length/demo.js:20]: Should we use percentages instead for all the values, here and below? Works better in a responsive scenario.

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks, the functionality and doclets are fine now, except my internal comment about setting to 0.

File: ts/Series/BoxPlot/BoxPlotSeries.ts [hunk 9]
```diff
@@ -87,6 +88,63 @@ class BoxPlotSeries extends ColumnSeries {
...
+                yPos: number
+            ): SVGPath.Segment[] => {
+                const halfLen = (
+                        (
+                            typeof xLen === 'string' &&
+                            xLen.endsWith('%')
+                        ) ?
+                            relativeLength(xLen, halfWidth) :
+                            Number(xLen) / 2
+                    ),
```
[the remark this answers is not in the document]
Comment (TorsteinHonsi) [reply ts/Series/BoxPlot/BoxPlotSeries.ts:116]: That's how `relativeLength` works. If a number is passed, the plain number is returned. If a percentage string is passed, the fraction of the second argument is returned. It's a common pattern in our options, therefore the utility function.

[the remark this answers is not in the document]
Comment (TorsteinHonsi) [reply ts/Series/BoxPlot/BoxPlotSeries.ts:116]: That way we don't have to do the ifs and elses, typechecks and endsWith for each time we use this pattern.

**Review by TorsteinHonsi — APPROVED** [review]
